### PR TITLE
Build manager binary inside a container

### PR DESCRIPTION
**Pull request by jiatongw**
**What this PR does / why we need it**:

Instead of building `manager` binary from local, it's better to build from a container. 
The build command in the changes is consistent with the Dockerfile in the root directory

/cc @figo 

**Special notes for your reviewer**:
N/A

**Release note**:

```release-note
NONE
```

### CODE REVIEW

**Comment by ncdc** [comment]
Do we want to remove the ability to use make to do a local (non-docker) build? Or should we have 2 targets: 1 for local, 1 for docker?

**Comment by ncdc** [comment]
Also, this doesn't currently work in prow/CI:
```
docker: Cannot connect to the Docker daemon at unix:///var/run/docker.sock. Is the docker daemon running?.
```

**Comment by vincepri** [comment]
The `make manager` and others targets are usually used in local development and CI. The CI configuration we use today doesn't allow, as @ncdc pointed out, to use docker in docker, so we'd have to change our configuration in `test-infra` to make this work.

That said, what do you think of adding a new target called `docker-build-manager` instead of changing the local one?

**Comment by jiatongw** [comment]
Hi @ncdc @vincepri , I feel like having two targets for building the same binary is kind of redundant. If we want to keep the local build, then I can close the PR.

**Comment by vincepri** [comment]
I don't mind having both, the local one is also multi-platform, while we're hardcoding `linux` in the suggested change

**Comment by jiatongw** [comment]
> I don't mind having both, the local one is also multi-platform, while we're hardcoding `linux` in the suggested change

OK. Then I can do: 
1. set a `GOOS` env to default as linux and pass to docker command. This is more flexible. 
2. how about name the new target `manager-in-docker` ?

**Comment by vincepri** [comment]
+1 on setting the GOOS variable. Regarding the name of the target, unless you have strong feelings about it, `docker-<something>` is consistent with the other docker targets

**Comment by jiatongw** [comment]
> +1 on setting the GOOS variable. Regarding the name of the target, unless you have strong feelings about it, `docker-<something>` is consistent with the other docker targets

Addressed.

File: Makefile [hunk 1]
```diff
@@ -58,6 +59,16 @@ test-go: ## Run tests
 manager: lint-full ## Build manager binary
 	go build -o bin/manager sigs.k8s.io/cluster-api/cmd/manager
 
+.PHONY: docker-build-manager
+docker-build-manager:
+	docker run --rm \
+	-v "${PWD}:/go/src/github.com/kubernetes-sigs/cluster-api" \
```
Comment (ncdc) [inline Makefile]:
```suggestion
	-v "${PWD}:/go/src/sigs.k8s.io/cluster-api" \
```

File: Makefile [hunk 2]
```diff
@@ -58,6 +59,16 @@ test-go: ## Run tests
...
+	-v "${GOPATH}/bin:/go/bin" \
+	-w "/go/src/github.com/kubernetes-sigs/cluster-api" \
```
Comment (ncdc) [inline Makefile]:
```suggestion
	-w "/go/src/sigs.k8s.io/cluster-api" \
```

File: Makefile [hunk 3]
```diff
@@ -58,6 +59,16 @@ test-go: ## Run tests
 manager: lint-full ## Build manager binary
 	go build -o bin/manager sigs.k8s.io/cluster-api/cmd/manager
 
+.PHONY: docker-build-manager
+docker-build-manager:
+	docker run --rm \
+	-v "${PWD}:/go/src/github.com/kubernetes-sigs/cluster-api" \
+	-v "${GOPATH}/bin:/go/bin" \
```
Comment (ncdc) [inline Makefile]: Since we're using go modules, we probably shouldn't assume `GOPATH` is set. I would probably make this `${PWD}/bin/manager`.

Comment (jiatongw) [reply Makefile]: All addressed.  Added `lint-full` in line 63
Please review again. Thanks!

**Pull request merged**